xcode/Subconscious/Shared/Services/DatabaseService.swift

```diff
@@ -114,69 +108,67 @@ final class DatabaseService {
         }
     }
 
-    /// Read database metadata from string key
-    func readMetadata(key: String) throws -> String {
+    func savepoint(_ savepoint: String) throws {
+        try self.database.savepoint(savepoint)
+    }
+
+    func release(_ savepoint: String) throws {
+        try self.database.release(savepoint)
+    }
+
+    func rollback(_ savepoint: String) throws {
+        try self.database.rollback(savepoint)
+    }
+
+    /// Geven a sphere did, read the last known version that the database has
+    /// synced to (if any).
+    func readSphereSyncInfo(sphereIdentity: Did) throws -> String? {
         guard self.state == .ready else {
             throw DatabaseServiceError.notReady
         }
         let rows = try database.execute(
-            sql: "SELECT value FROM database_metadata WHERE key = ?",
-            parameters: [.text(key)]
+            sql: "SELECT version FROM sphere_sync_info WHERE did = ?",
+            parameters: [.text(sphereIdentity.description)]
         )
-        guard let version = rows.first?.col(0)?.toString() else {
-            throw DatabaseServiceError.notFound(
-                "No value found for key \(key)"
-            )
-        }
+        let version = rows.first?.col(0)?.toString()
         return version
     }
 
-    /// Read metadata from type-safe metadata key
-    func readMetadata(key: DatabaseMetaKeys) throws -> String {
-        try readMetadata(key: key.rawValue)
-    }
-
     /// Write database metadata at string key
-    func writeMetadatadata(key: String, value: String) throws {
+    func writeSphereSyncInfo(sphereIdentity: Did, version: String) throws {
         guard self.state == .ready else {
             throw DatabaseServiceError.notReady
         }
         try database.execute(
             sql: """
-            INSERT OR REPLACE INTO database_metadata (
-                key,
-                value
+            INSERT OR REPLACE INTO sphere_sync_info (
+                did,
+                version
             )
             VALUES (?, ?)
             """,
-            parameters: [.text(key), .text(value)]
+            parameters: [.text(sphereIdentity.description), .text(version)]
         )
     }
```

Comment:
Clever change, very flexible for handling multiple spheres without much disruption.